rfcs/032-calm-deletions/README.md

```diff
@@ -0,0 +1,30 @@
+# RFC 032: Calm deletion watcher
+
+**Status:** Draft
+
+**Last updated:** 28/01/2021
+
+## Background
+
+We've realised that the pipeline does not pick up deleted Calm records - once they're deleted there is no trace of them in the Calm API, but they remain in the source VHS tables. Especially because a cleanup project at the end of last year was responsible for a very large number of deletions, this needs to be resolved.
+
+So far as we can tell, there is no way to find deleted records other than checking for their absence. This means that our solution needs to take some sort of polling-based approach: going through the source table and checking for the continued presence of each record in the Calm API.
+
+## Proposal
+
+- The deletion watcher will be an ECS task within the existing calm adapter cluster.
+- It will run as a scheduled task at a frequency to be agreed with the collections information team.
+- It will also be possible to run the task manually via the `RunTask` API. A script will be provided for this.
```

Comment:
I feel like specifying the exact API here is overkill, and also not quite what I had in mind.

I imagine this would be an SQS-backed autoscaling worker (like everything else we have), and triggered by a Lambda posting a message to an SNS topic. If you want to run it manually, you run the Lambda through the console. Similar to the snapshot generator.

Reply:
I don't think it needs to be that complex, though - ECS [provides](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/scheduled_tasks.html) task scheduling and we use `RunTask` elsewhere (eg for the training tasks for inference services, and previously for the reindexer).

SQS is useful when dynamic input is required, but that's not the case here. What are the advantages of Lambda + SQS + SNS in this case?

Reply:
> SQS is useful when dynamic input is required, but that's not the case here

Hmm. Are we sure about that?

If you make this an SQS-backed worker that takes a (CalmSourceRecord) and checks the freshness of that, then:

1. You can reuse the reindexer pretty much unmodified for scanning the table
2. It's not an issue if you get interrupted halfway through a scan
3. We can yank individual records with a specific reindex outside the usual schedule, without having to rescan the whole table (which might take hours or days)

I'm sure there's a bias towards doing another SQS-backed worker because we already run a bunch of those and have plenty of tooling for running them. It's a much more well-understood process than using RunTask.

(The Lambda is probably wrong for this, you're right.)

Reply:
Hmmm, yeah, I can see the advantages - the ability to update specific records is probably the most useful aspect of this approach.

At the same time 2 tasks + scaffolding for something this simple seems a bit like overkill! It would also _double_ the number of table reads required as the reindexer sends (and needs to send, because of SQS message sizes) only references to source data rows, rather than the whole records. Not sure (as in, genuinely can't decide) if we'd want to incur that penalty for these advantages...

Reply:
Using only the `RunTask` API provides limited guarantees (and auditing) that the task that is launched completes successfully.

Reply:
> At the same time 2 tasks + scaffolding for something this simple seems a bit like overkill! It would also double the number of table reads required as the reindexer sends (and needs to send, because of SQS message sizes) only references to source data rows, rather than the whole records. Not sure (as in, genuinely can't decide) if we'd want to incur that penalty for these advantages...

Aha! So this may be a source of the difference of opinion here – the reindexer _used_ to do that, but no longer!

This is an optimisation Alice suggested – the reindexer now sends the complete record, to save doing the double DynamoDB lookup in the transformers. If the service can tap that feed (which can be added with a config change in the Terraform), it gets a copy of every row in DynamoDB.

Reply:
Ah, that does change things then! I'll have a look at what that might look like

Reply:
This is the case class that gets passed by the adapter/reindexer to the transformers: https://github.com/wellcomecollection/catalogue/blob/master/common/source_model/src/main/scala/weco/catalogue/source_model/SourcePayload.scala

It's not quite the DynamoDB row – it papers over some naming weirdness in the tables, e.g. the use of the word "payload" when it's really just a "location".